Code/Framework/AzCore/AzCore/DOM/DomPatch.h

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/DOM/DomPath.h>
+#include <AzCore/DOM/DomValue.h>
+#include <AzCore/std/containers/deque.h>
+
+namespace AZ::Dom
+{
+    //! A patch operation that represents an atomic operation for mutating or validating a Value.
+    //! PatchOperations can be created with helper methods in Patch. /see Patch
+    class PatchOperation final
+    {
+    public:
+        using PatchOutcome = AZ::Outcome<void, AZStd::string>;
+
+        //! The operation to perform.
+        enum class Type
+        {
+            Add, //!< Inserts or replaces the value at DestinationPath with Value
+            Remove, //!< Removes the entry at DestinationPath
+            Replace, //!< Replaces the value at DestinationPath with Value
+            Copy, //!< Copies the contents of SourcePath to DestinationPath
+            Move, //!< Moves the contents of SourcePath to DestinationPath
+            Test //!< Ensures the contents of DestinationPath match Value or fails, performs no mutations
+        };
+
+        PatchOperation() = default;
+        PatchOperation(const PatchOperation&) = default;
+        PatchOperation(PatchOperation&&) = default;
+
+        PatchOperation(Path destionationPath, Type type, Value value);
+        PatchOperation(Path destionationPath, Type type, Path sourcePath);
+        PatchOperation(Path path, Type type);
+
+        static PatchOperation AddOperation(Path destinationPath, Value value);
+        static PatchOperation RemoveOperation(Path pathToRemove);
+        static PatchOperation ReplaceOperation(Path destinationPath, Value value);
+        static PatchOperation CopyOperation(Path destinationPath, Path sourcePath);
+        static PatchOperation MoveOperation(Path destinationPath, Path sourcePath);
+        static PatchOperation TestOperation(Path testPath, Value value);
+
+        PatchOperation& operator=(const PatchOperation&) = default;
+        PatchOperation& operator=(PatchOperation&&) = default;
+
+        bool operator==(const PatchOperation& rhs) const;
+        bool operator!=(const PatchOperation& rhs) const;
+
+        Type GetType() const;
+        void SetType(Type type);
+
+        const Path& GetDestinationPath() const;
+        void SetDestinationPath(Path path);
+
+        const Value& GetValue() const;
+        void SetValue(Value value);
+
+        const Path& GetSourcePath() const;
+        void SetSourcePath(Path path);
+
+        AZ::Outcome<Value, AZStd::string> Apply(Value rootElement) const;
+        PatchOutcome ApplyInPlace(Value& rootElement) const;
```

Comment:
nit: seems to be a helper and can be private for now?

Reply:
This is deliberate public API - you can use this to apply a patch to a DOM in place, which may be more performant because there's no copy/detach.